File: linkedlist/doublylinkedlist.go
package linkedlist

import (
	"errors"
	"fmt"
)

type DoublyLinkedList struct {
	head *Node
	tail *Node
	size int
}

// Size() returns current size of doubly linked list
func (dl DoublyLinkedList) Size() int {
	return dl.size
}

// IsEmpty() returns boolean evaluating whether list is empty or not based on its size value
func (dl DoublyLinkedList) IsEmpty() bool {
	return dl.size == 0
}

// Add() adds new Node with integer value to the list
// Either first node in list or at the end of the list
func (dl *DoublyLinkedList) Add(val int) {
	if dl.IsEmpty() {
		dl.AddFirst(val)
	} else {
		dl.AddLast(val)
	}
}

// AddFirst() adds new Node with integer value at the beginnig of list
func (dl *DoublyLinkedList) AddFirst(val int) {
	n := &Node{val: val}
	if dl.IsEmpty() {
		dl.head = n
		dl.tail = n
	} else {
		oldHead := dl.head
		dl.head = n
		dl.head.next = oldHead
		oldHead.prev = dl.head
	}
	dl.size += 1
}

func (dl *DoublyLinkedList) AddLast(val int) {
	prevTail := dl.tail
	dl.tail = &Node{val: val}
	prevTail.next = dl.tail
	dl.tail.prev = prevTail
	dl.size += 1
}

func (dl *DoublyLinkedList) AddAt(idx, val int) (bool, error) {
	if idx > dl.size || idx < 0 {
		return false, &ErrLinkedList{Err: errors.New("Index out of list bounds!")}
	}
	prev := dl.head
	for i, node := 0, dl.head; node != nil; i++ {
		if i == idx {
			newNode := &Node{val, prev, prev.prev}
			prev.next = newNode
			node.prev = newNode
			dl.size += 1
			return true, nil
		}
		prev = node
		node = node.next
	}
	return false, &ErrLinkedList{Err: errors.New("Couldn't add value at specified index!")}
}

func (dl *DoublyLinkedList) RemoveFirst() (int, error) {
	if dl.IsEmpty() {
		return -1, &ErrLinkedList{Err: errors.New("Couldn't remove item!"), Cause: "List is empty!"}
	}
	data := dl.head.Peek()
	newHead := dl.head.next
	dl.head = newHead
	dl.head.prev = nil
	dl.size -= 1
	return data, nil
}

func (dl *DoublyLinkedList) RemoveLast() (int, error) {
	if dl.IsEmpty() {
		return -1, &ErrLinkedList{Err: errors.New("Couldn't remove item!"), Cause: "List is empty!"}
	}
	data := dl.tail.Peek()
	newTail := dl.tail.prev
	dl.tail = newTail
	dl.tail.next = nil
	dl.size -= 1
	return data, nil
}

func (dl DoublyLinkedList) String() string {
	var ret string
	for i, node := 0, dl.head; node != nil; i++ {
		ret = ret + fmt.Sprintf("%d: %d\n", i, node.Peek())
		node = node.next
	}
	return fmt.Sprintf("%s", ret)
}

File: py/src/linked_list/doubly_linked_list.py
"""
Implementation of doubly linked lists in Python
"""

from typing import TypeVar

from src.linked_list.node import NextPrevNode

T = TypeVar("T")


class DoublyLinkedList[T]:
    def __init__(self, value: T) -> None:
        node = NextPrevNode(value)
        self._head: NextPrevNode[T] = node
        self._tail: NextPrevNode[T] = node

    @property
    def head(self) -> T:
        return self._head.value

    @property
    def tail(self) -> T:
        return self._tail.value

    def add_tail(self, value: T) -> None:
        node = NextPrevNode(value)
        tail_node = self._tail
        tail_node.next = node
        node.prev = tail_node
        self._tail = node

    def add_head(self, value: T) -> None:
        node = NextPrevNode(value)
        head_node = self._head
        head_node.prev = node
        node.next = head_node
        self._head = node

    def remove_tail(self) -> T:
        prev_tail = self._tail
        new_tail = prev_tail.prev
        new_tail.next = None
        self._tail = new_tail
        return prev_tail.value

    def remove_head(self) -> T:
        if self._head.next:
            prev_head = self._head
            new_head = prev_head.next
            new_head.prev = None
            self._head = new_head
            return prev_head.value
        prev_head = self._head
        self._head = None
        self._tail = None
        return prev_head.value

    def traverse(self) -> None:
        node = self._head
        while node.next != None:
            print("NODE: ", node.value)
            node = node.next
        print("NODE: ", node.value)

    def reverse(self) -> None:
        node = self._tail
        while node.prev != None:
            print("NODE: ", node.value)
            node = node.prev
        print("NODE: ", node.value)
